core/util/proxy.py

```diff
@@ -0,0 +1,108 @@
+"""
+OWASP Maryam!
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""
+
+#import requests
+from bs4 import BeautifulSoup
+import time
+#Python proxy module. Contributed by: @anuran-roy under @saeeddhqan
+class main:
+
+    def __init__(self):
+        self.framework.use = self.framework._global_options_['autoproxy']
+        self.framework.iplist = []
+
+    def getproxy(self):
+        #if self.use:
+        self.framework.output('[PROXY] Gathering proxies. It is quite time consuming, so do some pushups in the meantime :P')
+        purls = [ 'https://premproxy.com/list/' ]
+        i = 2
+        while i <= 8:
+            purl = f"https://premproxy.com/list/0{str(i)}.htm"
+            purls.append(purl)
+            i += 1    
+        ips = []
+        ips2 = []
+        purl = purls[0]
+        try:
+            for purl in purls:
+                r = self.framework.request(url=purl, method="GET")
+                sp = BeautifulSoup(r.content,'html.parser')
+                lst = list(sp.find_all('input',attrs = {"name" : "proxyIp[]"}))
+                ips = [str(l.get('value'))[0:-6] for l in lst]    
+                time.sleep(2)
+            #self.framework.output(ips[0:5])
+            self.framework.output('[PROXY] Proxies gathered.')
+        except Exception as e1:
+            self.framework.error(f'[PROXY] An error occured. Error details: {e1}')    
+        j = 0
+        while j<len(ips):
+            ips[j] = f"http://{ips[j]}"
+            j += 1
+        j = 0
+        self.framework.output('[PROXY] Now checking for availability of the proxies...')
+        while j<len(ips):
+            try:
+                if int(self.framework.request(url=ips[j],method="GET").status_code) == 200:
```

Comment:
This is a not a good way to check whether a proxy is up.
Because each connection with an invalid proxy will last for 10 seconds which is Maryam's default global timeout.
Instead use `self.framework.request(url=ips[j], method="GET", timeout=2)` to specify a timeout of two seconds.

Reply:
Done. I have set it timeout to 5 seconds, just in case of a bit slower connections. 😅

Reply:
Why I said 2 is because you are making compromises here either way.
By setting it to 5 instead; 
1. Yes you will maybe successfully find a proxy that works but is a little slower.
2. But 3 more seconds will be wasted on each invalid proxy.

But by setting it to 2;
1. You will not detect slower proxies, but you will not waste much time on invalid proxies
2. But you will eventually find a fast, good proxy.

I suggest you test with many timeout levels to see which performs best.

Reply:
Yes Sir, I'm on it 😃